parlai/core/torch_ranker_agent.py

```diff
@@ -0,0 +1,351 @@
+from parlai.core.utils import round_sigfigs, padded_tensor
+
+class TorchRankerAgent(TorchAgent):
+    @staticmethod
+    def add_cmdline_args(argparser):
+        TorchAgent.add_cmdline_args(argparser)
+        agent = argparser.add_argument_group('TorchRankerAgent')
+        agent.add_argument(
+            '-tc', '--train_candidates', type=str, default='batch',
+            choices=['batch', 'inline', 'fixed', 'dict'],
+            help='the source of candidates during training')
+        agent.add_argument(
+            '-ec', '--eval_candidates', type=str, default='batch',
+            choices=['batch', 'inline', 'fixed', 'dict'],
+            help='the source of candidates during training')
+        agent.add_argument(
+            '-cands', '--fixed-candidates-path', type=str,
+            help='a text file of fixed candidates to use for all examples, '
+                'one candidate per line')
+
+    def __init__(self, opt, shared):
+        if shared:
+            super().__init__(opt, shared)
+            self.model = shared['model']
+            self.metrics = shared['metrics']
+        else:
+            # Must call _get_model_file first so that paths are updated if
+            # necessary (e.g., .dict file)
+            model_file = self._get_model_file(opt)
+            self.metrics = {'loss': 0.0, 'examples': 0, 'rank': 0}
+            opt['rank_candidates'] = True
+
+            super().__init__(opt, shared)
+
+            print(f'Building model of type {self.id}')
+            self.build_model()
+            if model_file:
+                print('Loading existing model parameters from ' + model_file)
+                self.load(model_file)
+
+        self.rank_loss = nn.CrossEntropyLoss(reduction='sum')
+        self._set_fixed_candidates(shared)
+
+        if self.use_cuda:
+            self.model.cuda()
+            self.rank_loss.cuda()
+
+        optim_params = [p for p in self.model.parameters() if p.requires_grad]
+        self._init_optim(optim_params)
+
+    def score_candidates(self, batch, cand_vecs):
+        """Given a batch and candidate set, return scores (for ranking)"""
+        raise NotImplementedError(
+            'Abstract class: user must implement score()')
+
+    def build_model(self):
+        """Build a new model (implemented by children classes)"""
+        raise NotImplementedError(
+            'Abstract class: user must implement build_model()')
+
+    def train_step(self, batch):
+        """Train on a single batch of examples."""
+        if batch.text_vec is None:
+            return
+        batchsize = batch.text_vec.size(0)
+        self.model.train()
+        self.optimizer.zero_grad()
+
+        cands, cand_vecs, label_inds = self._build_candidates(batch,
+            source=self.opt['train_candidates'], mode='train')
+        scores = self.score_candidates(batch, cand_vecs)
+        loss = self.rank_loss(scores, label_inds)
+
+        self.metrics['loss'] += loss.item()
```

Comment:
Note: ParlAI automatically calculates hits@k metrics IFF label_candidates are provided. If we are simply ranking with the negatives as other elements in the batch (or with a separate set of candidates), we must calculate accuracy, etc. ourselves.

Reply:
Thank you for the note! I don't want to duplicate the metrics code, so I'll just make sure that we pass back the ranked candidates in a way that ParlAI will recognize.

Reply:
I believe you copied this directly from memnn--keep the code from there. calculating the mean rank is a nice useful metric which is a little different than hits@ and is calculable MUCH faster than hits@k

Reply:
Ok, I can revert back to the original memnn code. It seems unfortunate though that the ranking-related metrics are housed in the Metrics object that we don't have very direct access to. Code duplication is especially scary when it comes to metrics; maybe at sme point we can break the calculation of ranking-related metrics into a single helper method that both Metrics and TorchRankerAgent call.

Reply:
I see what you mean, but do note that Metrics is acting on text (slow but fair / consistent) and mean_rank is being calculated on tensors (subject to the model doing it correctly but much faster to calculate and so better during training)